File: backend/Modules/Ranking/RankingEnvioObserver.cs
using CodeMX.Api.Modules.Envios;

namespace CodeMX.Api.Modules.Ranking;

/// <summary>
/// Observador concreto (Observer): cuando Envíos publica un envío ACEPTADO, el módulo
/// Ranking reacciona registrando el acierto (que a su vez consulta los puntos a Retos).
/// </summary>
public class RankingEnvioObserver : IEnvioObserver
{
    private readonly IRankingApi _ranking;

    public RankingEnvioObserver(IRankingApi ranking) => _ranking = ranking;

    public Task OnEnvioAceptadoAsync(EnvioAceptadoEvent evento) =>
        _ranking.RegistrarAciertoAsync(evento.UsuarioId, evento.RetoId);
}

File: backend-java/src/main/java/mx/codemx/modules/ranking/RankingEnvioObserver.java
package mx.codemx.modules.ranking;

import mx.codemx.modules.envios.EnvioAceptadoEvent;
import mx.codemx.modules.envios.EnvioObserver;
import org.springframework.stereotype.Component;

@Component
public class RankingEnvioObserver implements EnvioObserver {

    private final RankingApi ranking;

    public RankingEnvioObserver(RankingApi ranking) {
        this.ranking = ranking;
    }

    @Override
    public void onEnvioAceptado(EnvioAceptadoEvent evento) {
        ranking.registrarAcierto(evento.usuarioId(), evento.retoId());
    }
}
